### PR TITLE
Tests for the extract and literature endpoints

### PR DESCRIPTION
**Pull request by mruwnik**
Almost done with the endpoint tests...

### CODE REVIEW

File: src/stampy_nlp/routes.py [hunk 1]
```diff
@@ -86,7 +86,7 @@ def literature_api():
     return jsonify(lit_search(query, top_k=top_k))
 
 
-@api.route('/extract', methods=['GET'])
+@api.route('/extract', methods=['GET', 'POST'])
```
Comment (mruwnik) [inline src/stampy_nlp/routes.py:89]: is POST used?

File: tests/conftest.py [hunk 2]
```diff
@@ -33,7 +33,24 @@ def stampy_duplicates():
...
 @pytest.fixture
-def mock_retriever_model():
+def mock_pinecone_search():
+    def get_pinecode_results(xq, namespace, top_k=None, filter=None, **kwargs):
+        """There are a load of Pinecone results saved as json results, so just read from the appropriate one."""
+        with open(Path(__file__).parent / (f'data/pinecone/{xq}+{namespace}+{filter}+{top_k}.json')) as f:
+            data = json.load(f)
+
+        matches = [
+            Mock(id=item['id'], score=item.get('score'), metadata=item.get('metadata', item), __getitem__=getattr)
```
Comment (mruwnik) [inline tests/conftest.py:43]: a bit of magic here, so the results can be used both as `result.field` and `result["field"]`

File: src/stampy_nlp/routes.py (hunk 1 above)
Comment (ccstan99) [reply src/stampy_nlp/routes.py:89]: Not right now. From what I gathered from recommended API design, it seems like POST & PUT were meant for updating & creating items, so maybe use them for adding encodings to pinecone?

File: tests/conftest.py (hunk 2 above)
Comment (ccstan99) [reply tests/conftest.py:44]: I noticed sometimes `result.field` was allowed but other times `result["field"]` was required. Must admit I didn't quite understand when or why.

**Review by ccstan99 — COMMENTED** [review]
Don't worry too much about setting up tests for the extract & literature endpoints yet. They're likely to change quite a bit in the upcoming weeks since they're still very early stage prototypes. I just wanted to make sure the semantic search is stable and someone else besides me is familiar with what's going on since it's being used in a few places between the bot, ui, and coda. But if you'd like to merge these changes, I'm ok with that too.

**Comment by mruwnik** [comment]
Hmm. I suppose it depends on how much it'll annoy you to have to keep fixing the tests :D

File: src/stampy_nlp/routes.py (hunk 1 above)
Comment (mruwnik) [reply src/stampy_nlp/routes.py:89]: cool. I was wondering whether you wanted to handle long queries, as GETs have a tendency to stop working on some browsers when the URL has more than 4000 characters

Comment (ccstan99) [reply src/stampy_nlp/routes.py:89]: That's actually what I was thinking when I originally supported POST requests. I took it out when I ran across several guides on good API design. The current endpoint naming convention doesn't really follow the good design guidelines either. The character limit probably isn't as big of an issue for basic Q&A, but may be a problem if we try to support chat conversations and need to keep track of context.

**Review by ccstan99 — COMMENTED** [review]
Yeah, does sound rather annoying, but I like that's you've set up framework. Does it make any sense to leave the tests for papers search & extract qa but not run/call them until that code is more finalized? Or should we just wait to update the code later? Since this'll likely come back on your plate, I'll give you the discretion on which workflow is easier for you.

**Comment by mruwnik** [comment]
I was thinking of refactoring some things, so I'd prefer to have them around for now. I'll add an extra flag. An ugly solution, but easier to clean up :/

File: tests/conftest.py [hunk 3]
```diff
@@ -43,38 +61,51 @@ def get_retriever_encodings(query):
...
+@pytest.fixture
+def mock_huggingface_search(mock_pinecone_search):
+    with patch('stampy_nlp.search.hf_search', lambda query: query):
+        yield
 
 
 # Add a command to run live tests - this will by default skip tests marked as live
 def pytest_addoption(parser):
+    parser.addoption(
+        "--rununstable", action="store_true", default=False, help="run tests for parts of the code that are unstable"
```
Comment (mruwnik) [inline tests/conftest.py:88]: running `pytest --rununstable` will also run the tests for extract and literature, otherwise they'll get skipped

**Review by ccstan99 — APPROVED** [review]
OK, having the option to skip them by default should be fine then. Thanks!

**Pull request merged**